src/ProjectSystemTools/BuildLogging/Model/Backend/BackendBuildTableDataSource.cs

```diff
@@ -0,0 +1,127 @@
+
+        public string Identifier => BuildTableDataSourceIdentifier;
+
+        public string DisplayName => BuildDataSourceDisplayName;
+
+        public bool SupportRoslynLogging => _roslynLogger.Supported;
+
+        public bool IsLogging { get; private set; }
+
+        public bool SupportsRoslynLogging => _roslynLogger.Supported;
+
+        private NotifyCallback NotifyUI { get; set; }
+
+        public BackendBuildTableDataSource()
+        {
+            _evaluationLogger = new EvaluationLogger(this);
+            _roslynLogger = new RoslynLogger(this);
+        }
+
+        public void Start(NotifyCallback notifyCallback)
+        {
+            NotifyUI = notifyCallback;
+
+            IsLogging = true;
+            ProjectCollection.GlobalProjectCollection.RegisterLogger(_evaluationLogger);
+            _roslynLogger.Start();
+        }
+
+        public void Stop()
+        {
+            NotifyUI = null;
+
+            IsLogging = false;
+            ProjectCollection.GlobalProjectCollection.UnregisterAllLoggers();
+            _roslynLogger.Stop();
+        }
+
+        public void Clear()
+        {
+            foreach (var build in _entries)
+            {
+                build.Dispose();
+            }
+            _entries = ImmutableList<Build>.Empty;
+        }
+
+        public ILogger CreateLogger(bool isDesignTime) => new ProjectLogger(this, isDesignTime);
+
+        /// <summary>
+        /// (Temporary) return log path on server for a given build
+        /// If buildID cannot be found, will return null
+        /// </summary>
+        /// <param name="buildID">ID to return build for</param>
+        /// <returns>(Temporary) returns filepath to log path (on server)</returns>
+        public string RetrieveLogForBuild(int buildID)
+        {
+            Build match = findBuildByID(buildID);
+            return match.LogPath;
+        }
+
+        /// <summary>
+        /// 
+        /// </summary>
+        /// <param name="buildID">ID to return build for</param>
+        /// <returns>Build matching the given ID, if no match, return null</returns>
+        private Build findBuildByID(int buildID)
+        {
+            return _entries.Find(x => x.BuildSummary.BuildID == buildID);
+        }
+
+        public ImmutableList<IBuildSummary> RetrieveAllBuilds()
+        {
+            ImmutableList<IBuildSummary> buildSummaries = ImmutableList<IBuildSummary>.Empty;
+            IEnumerator<Build> builds = _entries.GetEnumerator();
+            while (builds.MoveNext())
+            {
+                Build current = builds.Current;
+                buildSummaries = buildSummaries.Add(current.BuildSummary);
+            }
+            return buildSummaries;
+        }
+
+        public void Dispose()
+        {
+            Clear();
+        }
+
+        public void NotifyChange()
+        {
+            if (NotifyUI != null) 
+            {
+                NotifyUI();
+            }
```

Comment:
```suggestion
            NotifyUI?.Invoke();
```

Apart from being slightly less code, this fixes a potential race condition where the single subscriber unsubscribes between the null check and the invocation. It's very unlikely to happen, but thankfully easy to fix. Another fix would be to assign the handler to a local before null testing the local and invoking the local. That's essentially what `?.Invoke()` does.